File: app/src/main/java/com/innso/mobile/ui/viewModels/SplashViewModel.java
package com.innso.mobile.ui.viewModels;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.innso.mobile.BuildConfig;
import com.innso.mobile.api.controllers.AppControllerApi;
import com.innso.mobile.api.models.app.GeneralInformation;
import com.innso.mobile.ui.activities.LoginActivity;
import com.innso.mobile.ui.activities.MainActivity;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.subjects.PublishSubject;

public class SplashViewModel extends BaseViewModel {

    private PublishSubject<Boolean> updateVersion = PublishSubject.create();

    @Inject
    AppControllerApi appControllerApi;

    @Inject
    FirebaseAuth firebaseAuth;

    public SplashViewModel() {
        getComponent().inject(this);
    }


    public void validateAuthentication() {
        appControllerApi.checkVersion()
                .subscribe(this::validateVersion, this::showServiceError);
    }

    private void validateVersion(GeneralInformation generalInformation) {
        if (BuildConfig.VERSION_CODE >= generalInformation.getBuildVersion()) {
            FirebaseUser currentUser = firebaseAuth.getCurrentUser();
            if (currentUser != null && !currentUser.isAnonymous()) {
                startActivityEvent.onNext(MainActivity.class);
            } else {
                startActivityEvent.onNext(LoginActivity.class);
            }
        } else {
            updateVersion.onNext(true);
        }
    }

    public Observable<Boolean> updateVersionEvent() {
        return updateVersion.observeOn(AndroidSchedulers.mainThread());
    }

}

File: app/src/main/java/com/innso/mobile/viewModels/onBoarding/SplashViewModel.kt
package com.innso.mobile.viewModels.onBoarding

import android.app.Activity
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.google.firebase.auth.FirebaseAuth
import com.innso.mobile.BuildConfig
import com.innso.mobile.api.controllers.AppControllerApi
import com.innso.mobile.api.models.app.GeneralInformation
import com.innso.mobile.ui.activities.HomeActivity
import com.innso.mobile.ui.activities.MainActivity
import com.innso.mobile.viewModels.AndroidViewModel
import com.innso.mobile.viewModels.models.FinishActivityModel
import com.innso.mobile.viewModels.models.StartActivityModel
import javax.inject.Inject

class SplashViewModel : AndroidViewModel() {

    private val updateVersion = MutableLiveData<Boolean>()

    @Inject
    lateinit var appControllerApi: AppControllerApi

    @Inject
    lateinit var firebaseAuth: FirebaseAuth

    init {
        getComponent().inject(this)
    }

    fun validateAuthentication() {
        disposables.add(appControllerApi.checkVersion()
                .doOnSubscribe { showLoading() }
                .doFinally { hideLoading() }
                .subscribe({ this.validateVersion(it) }, { this.showServiceError(it) }))
    }

    private fun validateVersion(generalInformation: GeneralInformation) {
        if (BuildConfig.VERSION_CODE >= generalInformation.buildVersion) {
            val currentUser = firebaseAuth.currentUser
            if (currentUser != null && !currentUser.isAnonymous) {
                startActivity.postValue(StartActivityModel(MainActivity::class.java))
            } else {
                startActivity.postValue(StartActivityModel(HomeActivity::class.java))
            }
            closeView.postValue(FinishActivityModel(Activity.RESULT_OK))
        } else {
            updateVersion.postValue(true)
        }
    }

    fun updateVersionEvent(): LiveData<Boolean> = updateVersion

}
